### PR TITLE
feat: implement support for volatile events with new VolatileSocket w…

### PR DESCRIPTION
**Pull request by LiamCarPer**
## Motivation

Socket.io's Node.js implementation supports a `volatile` flag on emits that drops events when the underlying connection is not ready. This is highly useful for high-frequency, non-critical data like game position updates or telemetry, as it prevents buffer buildup on unstable connections.

This feature was requested in #602 and is documented in the [[Socket.io volatile events spec](https://www.google.com/search?q=https://socket.io/docs/v4/emitting-events/%23volatile-events)](https://www.google.com/search?q=https://socket.io/docs/v4/emitting-events/%23volatile-events). Currently, `socketioxide` has no equivalent, forcing users to either accept buffer pressure or implement their own messaging layer.

---

## Solution

Added a `Volatile` variant to `BroadcastFlags` (`0x04`) that flows through the existing adapter broadcast pipeline. On the local adapter, when `volatile` is set, errors from `send_many` are silently discarded rather than propagated, perfectly matching fire-and-forget semantics.

### User-Facing API

Three entry points are provided to mirror the Node.js patterns:

* **`Socket::volatile()`** Returns a `VolatileSocket<'a, A>` wrapper with a direct `emit()` that silently drops events when the socket is disconnected, the internal buffer is full, or encoding fails. The wrapper also exposes chain methods (`to`, `within`, `except`, `local`, `broadcast`, `timeout`) that delegate to `BroadcastOperators` with the volatile flag pre-set.
* **`BroadcastOperators::volatile()` and `ConfOperators::volatile()**` Sets the flag on the operator chain, enabling room-based broadcasting patterns like:
```rust
io.to("room").volatile().emit(...);

```

* **`SocketIo::volatile()`** A convenience alias on the default namespace for global emits:
```rust
io.volatile().emit(...);

```

### Adapter Propagation

The flag propagates directly through the `BroadcastOptions` struct. This ensures remote adapters (Redis, Postgres, MongoDB) receive it and can handle volatile semantics on their own nodes out of the box, without requiring any adapter-specific changes.

### CODE REVIEW

**Comment by Totodore** [comment]
Please check this comment regarding the implementation:
https://github.com/Totodore/socketioxide/issues/602#issuecomment-4676843802

you are missing a way to bypass the mpsc channel in engineio. Volatile packets will still be buffered, it is not what we want.

I might be wrong but, before Claude-ing something please try to understand the issue details before implementing something. Feel free to ask more details here or on the issue.

**Comment by LiamCarPer** [comment]
Thanks for the Review. You are right, my current approach still goes through the mpcs channel so volatile packets can buffer.
i am planning to add a separate  volatile send path in engineio, a second mpsc channel on Socket with capacity 1, paired with a send_volatile() method that does try _send iton it (drops if full). Then update both transporters to drain it with priority. Then thread the volatile flag BroadcastOptions through send_many so individual socket sends in broadcast flows also use the volatile engine.io path.

Does that sound good? Want to make sure i am in the right track before spending more time on it

**Comment by LiamCarPer** [comment]
Is this what you were looking for?

**Comment by Totodore** [comment]
Yes! One thing that we are missing with this solution is that volatile packets are out of order with classic packets. It might be ok though. We can simply document this specific case.
It would be nice to check what are the ordering constraints in the js implementation though and explore if there is any solution for this. I'll review this whenever I can.

**Comment by LiamCarPer** [comment]
Perfect, let me look into it

**Comment by LiamCarPer** [comment]
I checked js docs and they dont specify ordering constraints for volatile vs regular event.
I have added doc comments on both VolatileSocket and the engine.io emit.volatile() methods.

File: crates/engineioxide/src/transport/polling/mod.rs [hunk 1]
```diff
@@ -130,11 +130,28 @@ where
 
     let max_payload = engine.config.max_payload;
 
+    // Read the latest volatile packet from the watch channel, if any.
+    // Because the watch channel only retains the most recent value,
+    // any volatile messages that were overwritten since the last poll
+    // are automatically discarded.
+    let volatile_packets: Vec<_> = {
```
Comment (Totodore) [inline crates/engineioxide/src/transport/polling/mod.rs]: I don't think this is a correct approach, you are yielding the watch packet here. Meaning that if there is any .await point between now and the actual payload encoding and that in the meantime another volatile packet is sent by the user it will still be this one that is sent.

A better approach would be to refactor the mpsc receiver to a `Stream` and use [`.chain`](https://docs.rs/futures-util/latest/futures_util/stream/trait.StreamExt.html#method.zip). With this solution the payload encoding is not complexified.

I made this implementation in the current socketioxide client draft so you can use it as a base: https://github.com/Totodore/socketioxide/pull/555/changes#diff-e4fc706cfac0a21273c6a254bd1a3a2685541b63eca50bbd724afce0de9269a6R128

File: crates/engineioxide/Cargo.toml [hunk 2]
```diff
@@ -28,7 +28,7 @@ http-body.workspace = true
 serde.workspace = true
 serde_json.workspace = true
 thiserror.workspace = true
-tokio = { workspace = true, features = ["rt", "time"] }
+tokio = { workspace = true, features = ["rt", "time", "macros"] }
```
Comment (Totodore) [inline crates/engineioxide/Cargo.toml:31]: Do we need the macros feature?

File: crates/engineioxide/src/socket.rs [hunk 3]
```diff
@@ -488,6 +500,46 @@ where
...
     }
+
+    /// Try to send a volatile message bypassing the internal buffer channel.
+    /// Volatile messages may be dropped if the transport is not ready to
+    /// receive them.
+    ///
+    /// Because volatile messages bypass the main mpsc buffer queue, they may
+    /// arrive out of order relative to regular messages.
+    #[inline]
+    pub fn emit_volatile(&self, msg: impl Into<Str>) -> bool {
```
Comment (Totodore) [inline crates/engineioxide/src/socket.rs:514]: Pls document the return value.

File: crates/socketioxide/src/operators.rs [hunk 4]
```diff
@@ -228,6 +237,16 @@ impl<A: Adapter> BroadcastOperators<A> {
         self.timeout = Some(timeout);
         self
     }
+
+    /// Sets the volatile flag for the emit. When set, the event may be dropped
+    /// if the client is not ready to receive it (e.g. the connection is buffering or not connected).
+    /// This is useful for events that are not critical, such as position updates in a game.
+    ///
+    /// See [socket.io volatile events](https://socket.io/docs/v4/emitting-events/#volatile-events).
```
Comment (Totodore) [inline crates/socketioxide/src/operators.rs]: Could we move this to external doc to avoid duplication?
Like the other methods.

File: crates/socketioxide/src/socket.rs [hunk 5]
```diff
@@ -278,6 +278,90 @@ impl From<BroadcastError> for RemoteActionError {
...
+    pub fn broadcast(self) -> BroadcastOperators<A> {
+        self.socket.broadcast().volatile()
+    }
+
+    #[doc = include_str!("../docs/operators/timeout.md")]
+    pub fn timeout(self, timeout: Duration) -> BroadcastOperators<A> {
+        BroadcastOperators::from(ConfOperators::new(self.socket).timeout(timeout)).volatile()
+    }
+}
+
```
Comment (Totodore) [inline crates/socketioxide/src/socket.rs]: Whats the purpose of all of this?
Why not just having a `volatile` operator? Like the `broadcast` operator. Imagine if we would have implemented a new wrapper for each kind of send...

File: crates/socketioxide/src/socket.rs [hunk 6]
```diff
@@ -639,6 +723,29 @@ impl<A: Adapter> Socket<A> {
...
+    ///
+    /// let (_, io) = SocketIo::new_svc();
+    /// io.ns("/", async |socket: SocketRef| {
+    ///     socket.volatile().emit("position", &GameState { x: 1.0, y: 2.0 });
+    ///     socket.volatile().to("room1").emit("update", &42);
+    /// });
+    /// ```
+    pub fn volatile(&self) -> VolatileSocket<'_, A> {
+        VolatileSocket { socket: self }
+    }
```
Comment (Totodore) [inline crates/socketioxide/src/socket.rs:645]: Just return a `ConfOperator` like broadcast.

**Review by Totodore — CHANGES_REQUESTED** [review]
This is the good direction!
We need to rethink how the polling payload encoding is done though. You can pick some changes from #555 regarding the use of `Stream` in the encoder.

**Comment by LiamCarPer** [comment]
Hey man, sorry for being this late ran out of claude usage, jajjajaj just joking, was busy with work. 

I removed VolatileSocket, used ConfOperators with volatile flag, removed tokio macros feature, documented return values, moved volatile docs to external file. Test results, 79/79 unit tests, 106/109 doc tests, 0 clippy warnings, clean.

You had a concern about polling encoder, there are no .await points between the volatile snapshot and the encoder call, so the concern doesnt apply to the specific code,

Check and tell me is everything as you want it!

**Review by Totodore — COMMENTED** [review]
Hey, it is good, however we are missing integration tests for engineioxide and socketioxide  and unit tests for the encoder functions in engineioxide.
We need to test every combination possible of volatile vs non-volatile packets to ensure payloads are still correct.

**Comment by LiamCarPer** [comment]
I went ahead and added a bunch of comprehensive test coverage.

For the encoder unit tests in encoder.rs, I added 10 new tests checking the core logic for all three versions (v4, v3 string, and v3 binary). That covers volatile-only payloads, mixed normal and volatile ordering both ways, making sure the latest volatile data overwrites the old stuff, and testing what happens if volatile data drops in right while a buffer is draining.

Then for engineioxide integration tests in tests/volatile.rs, I added 3 tests to check the transport layer. It confirms volatile messages make it all the way through the full polling transport, double-checks the ordering when mixing volatile and normal messages, and makes sure those overwrite semantics hold up at the engine level too.

Lastly for socketioxide integration tests in tests/volatile.rs, i added 4 tests for the user-facing API and broadcasting. It ensures socket.volatile().emit() returns Ok(()), makes sure broadcasting with the volatile flag on doesn't panic, and checks that io.volatile() plays nice with the rest of our tools.

I have left two edge cases uncovered, Parked encoder + volatile arrives and Max payload boundary with volatile. What do you want me to do with it?

File: crates/engineioxide/src/transport/ws.rs [hunk 7]
```diff
@@ -288,18 +289,35 @@ async fn forward_to_socket<H: EngineIoHandler, S>(
...
             map_fn!(item);
         }
-        // For every available packet we continue to send until the channel is drained
         while let Ok(items) = internal_rx.try_recv() {
             for item in items {
                 map_fn!(item);
             }
         }
-
         tx.flush().await.ok();
```
Comment (Totodore) [inline crates/engineioxide/src/transport/ws.rs]: Just flush once inconditionnaly at the end, rather than flushing twice

**Review by Totodore — CHANGES_REQUESTED** [review]
Thanks for the tests, yes it would be nice to add parked encoder and max payload checks (especially because these are edge cases that may easily results in bugs).

**Comment by LiamCarPer** [comment]
I handled the WS flush consolidation in crates/engineioxide/src/transport/ws.rs. I removed that double-flush pattern where it was flushing once after the main channel drain and then again after volatile. Now there is just a single unconditional tx.flush().await.ok() at the end of each loop iteration, so it processes both the main and volatile packets before doing a single flush.

I also added 3 new parked encoder tests. These cover scenarios where volatile arrives while the encoder is blocked on recv_packet() and confirms that volatile isn't captured in that specific payload, only on the next poll. I explicitly tested this across v4, v3 string, and v3 binary encoders.

I added 3 new max payload volatile tests. These make sure that if volatile pushes data.len() over the limit, it correctly leaves the normal packets in the channel for the next poll. Like the others, this is fully tested for v4, v3 string, and v3 binary.

**Comment by Totodore** [comment]
I cant make it work on the whiteboard example, no events is being emitted. Try on your side to emit the drawing event with volatile and check if you get something. Same if I force polling transport, it doesn't work. Once it works with the whiteboard example check also with remote adapters (redis/postgres to find the issue).

Once you find the bug in the whiteboard example, please add regression tests.

Tip:
Maybe add more logs to track engineioxide behaviors, you can check with `RUST_LOG=socketioxide=trace,engineioxide=trace`.

**Comment by LiamCarPer** [comment]
Found it, volatile packets arriving while the encoder is parked on recv_packet() were not captured. The encoder checks volatile in the initial check and in the drain loop, but after parking, volatile arriving during that .await is missed, it only gets captured on the next poll.

I have added new regression tests "volatile_broadcast_arrives_via_polling_transport".

Thank you for the tip, it was useful

File: crates/engineioxide/src/transport/ws.rs [hunk 8]
```diff
@@ -288,17 +289,35 @@ async fn forward_to_socket<H: EngineIoHandler, S>(
             };
         }
 
-    while let Some(items) = internal_rx.recv().await {
+    loop {
+        // Priority: wait for and drain the main channel.
+        // Volatile events are checked after each main-channel cycle.
+        let items = match internal_rx.recv().await {
```
Comment (Totodore) [inline crates/engineioxide/src/transport/ws.rs]: This is still not working if you only send volatile events because `internal_rx` will wait indefinitely (until next ping...) here and block everything, you need to use [`tokio::select!`](https://docs.rs/tokio/latest/tokio/macro.select.html) to poll concurrently `internal_rx` and `volatile_rx`.

File: crates/engineioxide/src/transport/ws.rs [hunk 9]
```diff
@@ -288,17 +289,35 @@ async fn forward_to_socket<H: EngineIoHandler, S>(
...
         }
 
+        // Check volatile channel after main is drained.
+        // `has_changed` is non-blocking; if no volatile data is
+        // pending we simply go back to waiting for the main channel.
+        if volatile_rx.has_changed().unwrap_or(false) {
+            let val = volatile_rx.borrow_and_update().clone();
+            if let Some(packets) = val {
+                #[cfg(feature = "tracing")]
+                tracing::info!(sid = ?socket.id, "ws volatile check: flushing {:?}", &packets);
```
Comment (Totodore) [inline crates/engineioxide/src/transport/ws.rs]:
```suggestion
                tracing::debug!(sid = ?socket.id, "ws volatile check: flushing {} packets", packets.len());
```
We should only log in trace/debug and avoid printing the content of the packets (this would be extremely inefficient).

File: crates/socketioxide/tests/volatile.rs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            .volatile()
+            .emit("event", &"data")
+            .await
+            .ok();
+        tx.send(()).unwrap();
+    });
+
+    io.new_dummy_sock("/", ()).await;
+    rx.recv().unwrap();
+}
```
Comment (Totodore) [inline crates/socketioxide/tests/volatile.rs]:
```suggestion
```
Why would it panic in the first place? What are we testing here?
If there is no panic there is no reasons for this tests which asserts nothing.

File: crates/socketioxide/tests/volatile.rs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+async fn io_volatile_composes() {
+    let (_svc, io) = SocketIo::new_svc();
+    io.ns("/", |_: SocketRef| async {});
+
+    io.new_dummy_sock("/", ()).await;
+
+    // Volatile on the io handle delegates to the default namespace
+    let _ = io.volatile();
+    let _ = io.of("/").unwrap().volatile();
+}
```
Comment (Totodore) [inline crates/socketioxide/tests/volatile.rs]:
```suggestion
```

Same here, we want assertions, we dont need to test the API in itself.

File: crates/socketioxide/src/io.rs [hunk 12]
```diff
@@ -591,6 +591,16 @@ impl<A: Adapter> SocketIo<A> {
         self.get_default_op()
     }
 
+    /// _Alias for `io.of("/").unwrap().volatile()`_. If the **default namespace "/" is not found** this fn will panic!
+    ///
+    /// Sets the volatile flag on the broadcast operators. Volatile events may be dropped
+    /// if the client is not ready to receive them.
+    /// See [socket.io volatile events](https://socket.io/docs/v4/emitting-events/#volatile-events).
+    #[inline]
+    pub fn volatile(&self) -> BroadcastOperators<A> {
```
Comment (Totodore) [inline crates/socketioxide/src/io.rs:596]: We are missing a reference to the md doc (see other methods).

File: crates/socketioxide/src/operators.rs [hunk 13]
```diff
@@ -101,6 +112,18 @@ impl<A: Adapter> ConfOperators<'_, A> {
...
+                return Ok(());
+            }
+            let Ok(packet) = self.get_packet(event, data) else {
+                return Ok(());
+            };
+            self.socket
+                .send_raw_volatile(self.socket.parser.encode(packet));
+            return Ok(());
+        }
+
```
Comment (Totodore) [inline crates/socketioxide/src/operators.rs:126]: Would make more sense to move this after the if guard that checks if the socket is connected.

File: crates/socketioxide/src/socket.rs [hunk 14]
```diff
@@ -639,6 +639,27 @@ impl<A: Adapter> Socket<A> {
...
+    /// # use serde::Serialize;
+    /// #[derive(Serialize)]
+    /// struct GameState { x: f64, y: f64 }
+    ///
+    /// let (_, io) = SocketIo::new_svc();
+    /// io.ns("/", async |socket: SocketRef| {
+    ///     socket.volatile().emit("position", &GameState { x: 1.0, y: 2.0 }).ok();
+    ///     socket.volatile().to("room1").emit("update", &42).await.ok();
+    /// });
+    /// ```
```
Comment (Totodore) [inline crates/socketioxide/src/socket.rs]: Replace with the md doc.

File: crates/socketioxide-core/src/adapter/mod.rs [hunk 15]
```diff
@@ -202,6 +207,12 @@ pub trait SocketEmitter: Send + Sync + 'static {
     fn get_remote_sockets(&self, sids: BroadcastIter<'_>) -> Vec<RemoteSocketData>;
     /// Send data to the list of socket ids.
     fn send_many(&self, sids: BroadcastIter<'_>, data: Value) -> Result<(), Vec<SocketError>>;
+    /// Send data to the list of socket ids with volatile semantics.
+    /// Errors are silently discarded; packets may be dropped if the
+    /// transport is not ready.
+    fn send_many_volatile(&self, sids: BroadcastIter<'_>, data: Value) {
```
Comment (Totodore) [inline crates/socketioxide-core/src/adapter/mod.rs]: The default impl here is useless and really confusing.

File: crates/socketioxide/src/ns.rs [hunk 16]
```diff
@@ -268,6 +271,15 @@ impl<A: Adapter> InnerEmitter for Namespace<A> {
         if errs.is_empty() { Ok(()) } else { Err(errs) }
     }
 
+    fn send_many_volatile(&self, sids: BroadcastIter<'_>, data: Value) {
+        let sockets = self.sockets.read().unwrap();
+        for sid in sids {
+            if let Some(socket) = sockets.get(&sid) {
+                socket.send_raw_volatile(data.clone());
+            }
+        }
```
Comment (Totodore) [inline crates/socketioxide/src/ns.rs]: NIT: that's just to match with the functional pattern above.
```suggestion
        sids.filter_map(|sid| sockets.get(&sid))
            .for_each(|socket| socket.send_raw_volatile(data.clone()));
```

File: crates/socketioxide/tests/volatile.rs [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+#[tokio::test]
+async fn volatile_emit_returns_ok() {
+    use serde_json::json;
+    let (_svc, io) = SocketIo::new_svc();
+
+    let (tx, rx) = std::sync::mpsc::channel();
+    io.ns("/", async move |socket: SocketRef| {
+        let result = socket.volatile().emit("test", &json!({"key": "val"}));
+        tx.send(result).unwrap();
```
Comment (Totodore) [inline crates/socketioxide/tests/volatile.rs]: It would also be interesting to test that it returns false on the second call to volatile emit.

File: crates/socketioxide/tests/volatile.rs
Comment (Totodore) [inline crates/socketioxide/tests/volatile.rs:1]: Don't use std::mpsc, they are sync and we are in an an async context. This may block the testing harness or introduce flaky tests.

**Review by Totodore — CHANGES_REQUESTED** [review]
For WS transport the issue still appears, here are some tests below to reproduce the issue.
For polling transport this is even weirder when I test the whiteboard example I have nothing but then after the next ping I get an infinite loop of messages. Moreover all the payload encoder code is a mess mainly because we are using the wrong abstraction. What I'm going to do is to merge #555 so payload encoding uses streams. With stream it will be possible to merge gracefully the two channels without touching the encoder code.

In the meantime you can still fix the websocket code :).

```rust

#[tokio::test]
async fn volatile_only_arrives_via_polling() {
    let (socket_tx, mut socket_rx) = mpsc::channel(10);
    let mut svc = create_server(VolatileHandler { socket_tx }).await;
    let sid = create_polling_connection(&mut svc).await;

    let socket = socket_rx
        .recv()
        .await
        .expect("socket not received from on_connect");

    // Emit ONLY a volatile packet, nothing on the main channel.
    assert!(socket.emit_volatile("volatile_only"));

    let response = send_req(
        &mut svc,
        format!("transport=polling&sid={sid}"),
        http::Method::GET,
        None,
    )
    .await;
    assert_eq!(response, "volatile_only");
}

#[tokio::test]
async fn volatile_only_arrives_via_ws() {
    let (socket_tx, mut socket_rx) = mpsc::channel(10);
    let mut svc = create_server(VolatileHandler { socket_tx }).await;
    let mut stream = create_ws_connection(&mut svc).await;

    let socket = socket_rx
        .recv()
        .await
        .expect("socket not received from on_connect");

    let _open_packet = stream.next().await.unwrap().unwrap();

    let ping = stream.next().await.unwrap().unwrap();
    assert_eq!(ping, Message::Text("2".into()));

    assert!(socket.emit_volatile("volatile_only"));

    // The volatile must be delivered promptly, well before the next heartbeat
    // ping (300ms) would otherwise wake the main channel.
    let volatile = tokio::time::timeout(Duration::from_millis(100), stream.next()).await;

    let msg = volatile
        .expect("timeout: volatile-only packet was never flushed over websocket")
        .unwrap()
        .unwrap();
    assert_eq!(msg, Message::Text("4volatile_only".into()));
}
```

**Comment by LiamCarPer** [comment]
Hey, took care of all the feedback:
Fixed the WS transport bug by switching to tokio::select! so volatile-only events don't get blocked waiting on the next ping.

Cleaned up the tests (swapped to tokio mpsc, added the overwrite test, and included your two new integration tests).

Moved the docs for io.volatile() and socket.volatile() to the md include.

Knocked out all the nits (ns.rs refactor, tracing level, default impl removal, etc.).

Everything is green now

**Comment by Totodore** [comment]
Hey @LiamCarPer, I took the liberty to fix some things myself on top of your PR. So https://github.com/Totodore/socketioxide/pull/753 will likely superseed this one. In any case you will be in the contributors list.
Don't hesitate to check it.
Again thanks for the contribution!
